src/v/compression/lz4_decompression_buffers.cc

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "compression/lz4_decompression_buffers.h"
+
+#include "base/vassert.h"
+
+#include <seastar/coroutine/all.hh>
+
+namespace compression {
+
+std::ostream& operator<<(
+  std::ostream& os, lz4_decompression_buffers::alloc_ctx::allocation_state st) {
+    switch (st) {
+        using enum compression::lz4_decompression_buffers::alloc_ctx::
+          allocation_state;
+    case no_buffers_allocated:
+        return os << "no buffers allocated";
+    case input_buffer_allocated:
+        return os << "input buffer allocated";
+    case output_buffer_allocated:
+        return os << "output buffer allocated";
+    case both_buffers_allocated:
+        return os << "both buffers allocated";
+    }
+}
+
+lz4_decompression_buffers::lz4_decompression_buffers(
+  size_t buffer_size, size_t min_alloc_threshold, bool disabled)
+  : _min_alloc_threshold{min_alloc_threshold}
+  , _disabled{disabled} {
+    if (!_disabled) {
+        _buffers = {
+          .input_buffer = ss::allocate_aligned_buffer<char>(buffer_size, 8),
+          .output_buffer = ss::allocate_aligned_buffer<char>(buffer_size, 8),
+          .state = alloc_ctx::allocation_state::no_buffers_allocated,
+        };
+    }
+}
+
+bool lz4_decompression_buffers::has_reserved_buffers() const {
+    return _reservation.current() == 0;
+}
+
+ss::future<std::optional<ssx::semaphore_units>>
+lz4_decompression_buffers::reserve_buffers() {
+    if (_disabled) {
+        co_return std::nullopt;
+    }
+
+    co_return co_await ss::get_units(_reservation, 1);
```

Comment:
does nullopt communicate a special state? you could also return the reservation anyway since its just incrementing an integer by 1 and then not have any special cases / double the states for the return value.

Reply:
I wanted a fast path when the system was disabled so one fiber does not have to wait for another if the preallocation is turned off completely, although like discussed elsewhere on this PR I think no fiber will ever need to wait to get these units anyway, since there are no suspension points after this reserve call in any of our code paths.

Reply:
yeh, if there is no contention, then there is no overhead and no need for a fast path. also if there no contention, there is no need for a lock?

Reply:
removed the semaphore and related code